### ISSUE TITLE
SqlDataReader.ReadAsync() behaves synchronously sometimes (connect #3139210)

### ISSUE THREAD

**Issue by binki** [microsoft/dotnet#579]
I am cross posting this here because I haven’t gotten any response on https://connect.microsoft.com/VisualStudio/feedback/details/3139210 or https://twitter.com/ohnobinki/status/938524730340110337 or https://twitter.com/ohnobinki/status/916378365380132864 or https://stackoverflow.com/q/45597995/429091 and this appears like it might be the right place to submit this sort of stuff. The following is copy/paste+markup:

When looping over result sets and rows with `SqlClient`, sometimes `SqlDataReader.ReadAsync()` runs synchronously, blocking the calling thread for at least as long as the execution of the query prior to the next result set being received from SQL Server. This makes it unsafe to call from a GUI thread because doing so will result in the window being nonresponsive and marked “(Not Responding)”.

To reproduce, send a query which returns 1 result set, spends 20 seconds doing “busy work” (`WAITFOR DELAY '00:00:20'` does not reproduce the issue), and then returns another result set. Try to read that query from the GUI thread of a winforms application using `SqlCommand.ExecuteReaderAsync()`, `SqlDataReader.ReadAsync()`, and `SqlDataReader.NextResultAsync()`. Observe that the winforms application becomes “(Not Responding)” for almost 20 seconds.

The purpose of the `*Async()` calls in the framework is to enable thread-free GUI-thread-safe programming by using nonblocking I/O. If `SqlDataReader.ReadAsync()` runs synchronously and blocks the caller for indefinite amounts of time in some circumstances, that necessitates calling it through `Task.Run()` to avoid giving the user of the window a bad experience which defeats the whole point of even naming the method `ReadAsync()` in the first place.

I expect the query to complete after 20 seconds without causing the winforms application to become unresponsive.

I will upload a repro project. The repro project will require you to edit the connection strings according to your SQL Server configuration. I have also posted this at https://stackoverflow.com/q/45597995/429091, but I’ve decided this is likely a bug in the framework rather than in my usage of `SqlClient` so am filing here. In case it matters, I’m using .net-4.7.02542 and SQL Server 13.0.4411 on Windows 1703 build 16257.1.

Since the attachment never seems to show up after trying two times, I have uploaded the repro as a git repository to GitHub: https://github.com/binki/connect3139210

New info:

I can still repro this on .net-4.7.03016 (releasekey=461768) on Windows 1709 Build 17035.1000.

**Comment by binki** [comment]
After filing a “Technical Support” incident through a VS subscription, @Microsoft has acknowledged this bug and stated that it will be fixed in .net-4.7.3.

**Comment by binki** [comment]
I have verified that this is fixed in .net-4.8. You must target your app to .net-4.8 to see the improvement. I have not seen any documented `<AppContextOverride/>`—would be nice to have one to opt into the new functionality without having to target .net-4.8 (if I have to support old frameworks, would be still nice to progressively improve performance on newer ones). I have yet to see any evidence that .net-4.7.3 will exist.

I also see that this was addressed in corefx which was tracked at https://github.com/dotnet/corefx/issues/26594.
